src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -79,60 +93,78 @@ void WriteBufferFromAzureBlobStorage::finalizeImpl()
 {
     execWithRetry([this](){ next(); }, max_unexpected_write_error_retries);
 
-    if (tmp_buffer_write_offset > 0)
-        uploadBlock(tmp_buffer->data(), tmp_buffer_write_offset);
+    task_tracker->waitAll();
 
     auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
     execWithRetry([&](){ block_blob_client.CommitBlockList(block_ids); }, max_unexpected_write_error_retries);
 
     LOG_TRACE(log, "Committed {} blocks for blob `{}`", block_ids.size(), blob_path);
 }
 
-void WriteBufferFromAzureBlobStorage::uploadBlock(const char * data, size_t size)
+void WriteBufferFromAzureBlobStorage::nextImpl()
 {
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-    const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
+    task_tracker->waitIfAny();
 
-    Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(data), size);
-    execWithRetry([&](){ block_blob_client.StageBlock(block_id, memory_stream); }, max_unexpected_write_error_retries, size);
-    tmp_buffer_write_offset = 0;
+    reallocateBuffer();
+    detachBuffer();
 
-    LOG_TRACE(log, "Staged block (id: {}) of size {} (blob path: {}).", block_id, size, blob_path);
+    while (!detached_part_data.empty())
+    {
+        writePart(std::move(detached_part_data.front()));
+        detached_part_data.pop_front();
+    }
+
+    allocateBuffer();
 }
 
-WriteBufferFromAzureBlobStorage::MemoryBufferPtr WriteBufferFromAzureBlobStorage::allocateBuffer() const
+void WriteBufferFromAzureBlobStorage::allocateBuffer()
 {
-    return std::make_unique<Memory<>>(max_single_part_upload_size);
+    memory = Memory(max_single_part_upload_size);
+    WriteBuffer::set(memory.data(), memory.size());
```

Comment:
This function is not optimal. I mean `max_single_part_upload_size` is 100MB by default, and it's bad to always consume that much memory even to write a very small file. It's better to gradually increase the size of an allocated buffer with increasing the number of that buffer. Or maybe it's also good to always use the same buffer size. Please look at how `IBufferAllocationPolicy` is used in `WriteBufferFromS3`.